File: coq/plugin/features.ml

module Tag = struct
  type t =
    (* Types *)
    | Category
    | Object
    | Morphism
    | Functor
    | Equality
    (* Objects *)
    | AppliedFunctObj
    (* Morphisms *)
    | Identity
    | ComposeMph
    | AppliedFunctMph
    (* Equality *)
    | Reflexivity
    | Concat
    | InverseEq
    | ComposeEq
    | Associativity
    | LeftUnitality
    | RightUnitality
    | LeftApplication
    | RightApplication
    | FunctIdentity
    | FunctComposition
    | AppliedFunctEq

  let parse str =
    match str with
    (* Types *)
    | "category" -> Some Category
    | "object" -> Some Object
    | "morphism" -> Some Morphism
    | "functor" -> Some Functor
    | "equality" -> Some Equality
    (* Objects *)
    | "applied_funct_obj" -> Some AppliedFunctObj
    (* Morphisms *)
    | "identity" -> Some Identity
    | "compose_mph" -> Some ComposeMph
    | "applied_funct_mph" -> Some AppliedFunctMph
    (* Equality *)
    | "reflexivity" -> Some Reflexivity
    | "concat" -> Some Concat
    | "inverse_eq" -> Some InverseEq
    | "compose_eq" -> Some ComposeEq
    | "associativity" -> Some Associativity
    | "left_unitality" -> Some LeftUnitality
    | "right_unitality" -> Some RightUnitality
    | "left_application" -> Some LeftApplication
    | "right_application" -> Some RightApplication
    | "funct_identity" -> Some FunctIdentity
    | "funct_composition" -> Some FunctComposition
    | "applied_funct_eq" -> Some AppliedFunctEq
    | _ -> None
  let to_string = function
    (* Types *)
    | Category -> "category"
    | Object -> "object"
    | Morphism -> "morphism"
    | Functor -> "functor"
    | Equality -> "equality"
    (* Objects *)
    | AppliedFunctObj -> "applied_funct_obj"
    (* Morphisms *)
    | Identity -> "identity"
    | ComposeMph -> "compose_mph"
    | AppliedFunctMph -> "applied_funct_mph"
    (* Equality *)
    | Reflexivity -> "reflexivity"
    | Concat -> "concat"
    | InverseEq -> "inverse_eq"
    | ComposeEq -> "compose_eq"
    | Associativity -> "associativity"
    | LeftUnitality -> "left_unitality"
    | RightUnitality -> "right_unitality"
    | LeftApplication -> "left_application"
    | RightApplication -> "right_application"
    | FunctIdentity -> "funct_identity"
    | FunctComposition -> "funct_composition"
    | AppliedFunctEq -> "applied_funct_eq"

  module Eq = struct
    type nonrec t = t

    let tag_id = function
    | Category -> 0
    | Object -> 1
    | Morphism -> 2
    | Functor -> 3
    | Equality -> 4
    (* Prop ? *)
    | AppliedFunctObj -> 6
    | Identity -> 7
    | ComposeMph -> 8
    | AppliedFunctMph -> 9
    | Reflexivity -> 10
    | Concat -> 11
    | InverseEq -> 12
    | ComposeEq -> 13
    | Associativity -> 14
    | LeftUnitality -> 15
    | RightUnitality -> 16
    | LeftApplication -> 17
    | RightApplication -> 18
    | FunctIdentity -> 19
    | FunctComposition -> 20
    | AppliedFunctEq -> 21

    let compare tag1 tag2 = Int.compare (tag_id tag1) (tag_id tag2)
  end
end

type obj = Hyps.obj
type t =
  (* Types *)
  | Category
  | Object of (*cat*)obj
  | Morphism of (*cat*)obj * (*src*)obj * (*dst*)obj
  | Functor of (*src*)obj * (*dst*)obj
  | Equality of (*cat*)obj * (*src*)obj * (*dst*)obj * (*left*)obj * (*right*)obj
  (* Objects *)
  | AppliedFunctObj of (*src*)obj * (*dst*)obj * (*funct*)obj * (*obj*)obj
  (* Morphisms *)
  | Identity of (*cat*)obj * (*obj*)obj
  | ComposeMph of (*cat*)obj * (*src*)obj * (*mid*)obj * (*dst*)obj * (*m1*)obj * (*m2*)obj
  | AppliedFunctMph of (*src*)obj * (*dst*)obj * (*funct*)obj
                     * (*src*)obj * (*dst*)obj * (*mph*)obj
  (* Equality *)
  | Reflexivity of (*cat*)obj * (*src*)obj * (*dst*)obj * (*mph*)obj
  | Concat of (*cat*)obj * (*src*)obj * (*dst*)obj * (*left*)obj * (*mid*)obj * (*right*)obj
            * (*eq1*)obj * (*eq2*)obj
  | InverseEq of (*cat*)obj * (*src*)obj * (*dst*)obj * (*left*)obj * (*right*)obj * (*eq*)obj
  | ComposeEq of (*cat*)obj * (*src*)obj * (*mid*)obj * (*dst*)obj
               * (*left1*)obj * (*right1*)obj * (*eq1*)obj
               * (*left2*)obj * (*right2*)obj * (*eq2*)obj
  | Associativity of (*cat*)obj * (*src*)obj * (*mid1*)obj * (*mid2*)obj * (*dst*)obj
                   * (*m1*)obj * (*m2*)obj * (*m3*)obj
  | LeftUnitality of (*cat*)obj * (*src*)obj * (*dst*)obj * (*mph*)obj
  | RightUnitality of (*cat*)obj * (*src*)obj * (*dst*)obj * (*mph*)obj
  | LeftApplication of (*cat*)obj * (*src*)obj * (*mid*)obj * (*dst*)obj
                     * (*mph*)obj * (*left*)obj * (*right*)obj * (*eq*)obj
  | RightApplication of (*cat*)obj * (*src*)obj * (*mid*)obj * (*dst*)obj
                      * (*left*)obj * (*right*)obj * (*eq*)obj * (*mph*)obj
  | FunctIdentity of (*src*)obj * (*dst*)obj * (*funct*)obj * (*obj*)obj
  | FunctComposition of (*src*)obj * (*dst*)obj * (*funct*)obj
                      * (*src*)obj * (*mid*)obj * (*dst*)obj * (*m1*)obj * (*m2*)obj
  | AppliedFunctEq of (*src*)obj * (*dst*)obj * (*funct*)obj
                    * (*src*)obj * (*dst*)obj * (*left*)obj * (*right*)obj * (*eq*)obj

let tag = function
| Category -> Tag.Category
| Object _ -> Tag.Object
| Morphism _ -> Tag.Morphism
| Functor _ -> Tag.Functor
| Equality _ -> Tag.Equality
| AppliedFunctObj _ -> Tag.AppliedFunctObj
| Identity _ -> Tag.Identity
| ComposeMph _ -> Tag.ComposeMph
| AppliedFunctMph _ -> Tag.AppliedFunctMph
| Reflexivity _ -> Tag.Reflexivity
| Concat _ -> Tag.Concat
| InverseEq _ -> Tag.InverseEq
| ComposeEq _ -> Tag.ComposeEq
| Associativity _ -> Tag.Associativity
| LeftUnitality _ -> Tag.LeftUnitality
| RightUnitality _ -> Tag.RightUnitality
| LeftApplication _ -> Tag.LeftApplication
| RightApplication _ -> Tag.RightApplication
| FunctIdentity _ -> Tag.FunctIdentity
| FunctComposition _ -> Tag.FunctComposition
| AppliedFunctEq _ -> Tag.AppliedFunctEq

let to_list = function
| Category -> []
| Object cat -> [cat]
| Morphism (cat,src,dst) -> [cat;src;dst]
| Functor (src,dst) -> [src;dst]
| Equality (cat,src,dst,left,right) -> [cat;src;dst;left;right]
| AppliedFunctObj (src,dst,funct,obj) -> [src;dst;funct;obj]
| Identity (cat,obj) -> [cat;obj]
| ComposeMph (cat,src,mid,dst,m1,m2) -> [cat;src;mid;dst;m1;m2]
| AppliedFunctMph (src,dst,funct,srcm,dstm,mph) -> [src;dst;funct;srcm;dstm;mph]
| Reflexivity (cat,src,dst,mph) -> [cat;src;dst;mph]
| Concat (cat,src,dst,left,mid,right,eq1,eq2) ->
    [cat;src;dst;left;mid;right;eq1;eq2]
| InverseEq (cat,src,dst,left,right,eq) -> [cat;src;dst;left;right;eq]
| ComposeEq (cat,src,mid,dst,left1,right1,eq1,left2,right2,eq2) -> 
    [cat;src;mid;dst;left1;right1;eq1;left2;right2;eq2]
| Associativity (cat,src,mid1,mid2,dst,m1,m2,m3) ->
    [cat;src;mid1;mid2;dst;m1;m2;m3]
| LeftUnitality (cat,src,dst,mph) -> [cat;src;dst;mph]
| RightUnitality (cat,src,dst,mph) -> [cat;src;dst;mph]
| LeftApplication (cat,src,mid,dst,mph,left,right,eq) ->
    [cat;src;mid;dst;mph;left;right;eq]
| RightApplication (cat,src,mid,dst,left,right,eq,mph) ->
    [cat;src;mid;dst;left;right;eq;mph]
| FunctIdentity (src,dst,funct,obj) -> [src;dst;funct;obj]
| FunctComposition (src,dst,funct,srcm,midm,dstm,m1,m2) -> 
    [src;dst;funct;srcm;midm;dstm;m1;m2]
| AppliedFunctEq (src,dst,funct,srcm,dstm,left,right,eq) ->
    [src;dst;funct;srcm;dstm;left;right;eq]

let from_list tag args =
  let some x = Some x in
  match tag, args with
  (* Types *)
  | Tag.Category, [] -> Category |> some
  | Tag.Object, [cat] -> Object cat |> some
  | Tag.Morphism, [cat;src;dst] -> Morphism (cat,src,dst) |> some
  | Tag.Functor, [src;dst] -> Functor (src,dst) |> some
  | Tag.Equality, [cat;src;dst;left;right] -> Equality (cat,src,dst,left,right) |> some
  (* Objects *)
  | Tag.AppliedFunctObj, [src;dst;funct;obj] -> AppliedFunctObj (src,dst,funct,obj) |> some
  (* Morphisms *)
  | Tag.Identity, [cat;obj] -> Identity (cat,obj) |> some
  | Tag.ComposeMph, [cat;src;mid;dst;m1;m2] -> ComposeMph (cat,src,mid,dst,m1,m2) |> some
  | Tag.AppliedFunctMph, [scat;dcat;funct;src;dst;mph] ->
      AppliedFunctMph (scat,dcat,funct,src,dst,mph) |> some
  (* Equality *)
  | Tag.Reflexivity, [cat;src;dst;mph] -> Reflexivity (cat,src,dst,mph) |> some
  | Tag.Concat, [cat;src;dst;left;mid;right;eq1;eq2] ->
      Concat (cat,src,dst,left,mid,right,eq1,eq2) |> some
  | Tag.InverseEq, [cat;src;dst;left;right;eq] -> InverseEq (cat,src,dst,left,right,eq) |> some
  | Tag.ComposeEq, [cat;src;mid;dst;left1;right1;eq1;left2;right2;eq2] ->
      ComposeEq (cat,src,mid,dst,left1,right1,eq1,left2,right2,eq2) |> some
  | Tag.Associativity, [cat;src;mid1;mid2;dst;m1;m2;m3] ->
      Associativity (cat,src,mid1,mid2,dst,m1,m2,m3) |> some
  | Tag.LeftUnitality, [cat;src;dst;mph] ->
      LeftUnitality (cat,src,dst,mph) |> some
  | Tag.RightUnitality, [cat;src;dst;mph] ->
      RightUnitality (cat,src,dst,mph) |> some
  | Tag.LeftApplication, [cat;src;mid;dst;mph;left;right;eq] ->
      LeftApplication (cat,src,mid,dst,mph,left,right,eq) |> some
  | Tag.RightApplication, [cat;src;mid;dst;left;right;eq;mph] -> 
      RightApplication (cat,src,mid,dst,left,right,eq,mph) |> some
  | Tag.FunctIdentity, [src;dst;funct;obj] -> FunctIdentity (src,dst,funct,obj) |> some
  | Tag.FunctComposition, [scat;dcat;funct;src;mid;dst;m1;m2] ->
      FunctComposition (scat,dcat,funct,src,mid,dst,m1,m2) |> some
  | Tag.AppliedFunctEq, [scat;dcat;funct;src;dst;left;right;eq] ->
      AppliedFunctEq (scat,dcat,funct,src,dst,left,right,eq) |> some
  | _ -> None

module Eq = struct
  type nonrec t = t
  let compare x1 x2 =
    let cmp = Tag.Eq.compare (tag x1) (tag x2) in
    if cmp = 0
    then List.compare Hyps.compare_obj (to_list x1) (to_list x2)
    else cmp
end

File: engine/src/data/features.rs
use itertools::Itertools;

#[derive(Debug, Hash, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Tag {
    // Types
    Category,
    Object,
    Morphism,
    Functor,
    Equality,
    // Objects
    AppliedFunctObj,
    // Morphisms
    Identity,
    ComposeMph,
    AppliedFunctMph,
    // Equality
    Reflexivity,
    Concat,
    InverseEq,
    ComposeEq,
    Associativity,
    LeftUnitality,
    RightUnitality,
    LeftApplication,
    RightApplication,
    FunctIdentity,
    FunctComposition,
    AppliedFunctEq,
}

#[derive(Debug, Hash, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Feature {
    // Types
    Category,
    Object {
        cat: u64,
    },
    Morphism {
        cat: u64,
        src: u64,
        dst: u64,
    },
    Functor {
        scat: u64,
        dcat: u64,
    },
    Equality {
        cat: u64,
        src: u64,
        dst: u64,
        left: u64,
        right: u64,
    },
    // Objects
    AppliedFunctObj {
        scat: u64,
        dcat: u64,
        funct: u64,
        obj: u64,
    },
    // Morphisms
    Identity {
        cat: u64,
        obj: u64,
    },
    ComposeMph {
        cat: u64,
        src: u64,
        mid: u64,
        dst: u64,
        m1: u64,
        m2: u64,
    },
    AppliedFunctMph {
        scat: u64,
        dcat: u64,
        funct: u64,
        src: u64,
        dst: u64,
        mph: u64,
    },
    // Equality
    Reflexivity {
        cat: u64,
        src: u64,
        dst: u64,
        mph: u64,
    },
    Concat {
        cat: u64,
        src: u64,
        dst: u64,
        left: u64,
        mid: u64,
        right: u64,
        eq1: u64,
        eq2: u64,
    },
    InverseEq {
        cat: u64,
        src: u64,
        dst: u64,
        left: u64,
        right: u64,
        eq: u64,
    },
    ComposeEq {
        cat: u64,
        src: u64,
        mid: u64,
        dst: u64,
        left1: u64,
        right1: u64,
        eq1: u64,
        left2: u64,
        right2: u64,
        eq2: u64,
    },
    Associativity {
        cat: u64,
        src: u64,
        mid1: u64,
        mid2: u64,
        dst: u64,
        m1: u64,
        m2: u64,
        m3: u64,
    },
    LeftUnitality {
        cat: u64,
        src: u64,
        dst: u64,
        mph: u64,
    },
    RightUnitality {
        cat: u64,
        src: u64,
        dst: u64,
        mph: u64,
    },
    LeftApplication {
        cat: u64,
        src: u64,
        mid: u64,
        dst: u64,
        mph: u64,
        left: u64,
        right: u64,
        eq: u64,
    },
    RightApplication {
        cat: u64,
        src: u64,
        mid: u64,
        dst: u64,
        left: u64,
        right: u64,
        eq: u64,
        mph: u64,
    },
    FunctIdentity {
        scat: u64,
        dcat: u64,
        funct: u64,
        obj: u64,
    },
    FunctComposition {
        scat: u64,
        dcat: u64,
        funct: u64,
        src: u64,
        mid: u64,
        dst: u64,
        m1: u64,
        m2: u64,
    },
    AppliedFunctEq {
        scat: u64,
        dcat: u64,
        funct: u64,
        src: u64,
        dst: u64,
        left: u64,
        right: u64,
        eq: u64,
    },
}

pub struct FeatureIterator<'a> {
    feat: &'a Feature,
    pos: usize,
}

impl<'a> Iterator for FeatureIterator<'a> {
    type Item = u64;

    fn next(&mut self) -> Option<u64> {
        use Feature::*;
        let r = match (self.feat, self.pos) {
            (Category, _) => None,
            (Object { cat }, 0) => Some(cat),
            (Object { .. }, _) => None,
            (Morphism { cat, .. }, 0) => Some(cat),
            (Morphism { src, .. }, 1) => Some(src),
            (Morphism { dst, .. }, 2) => Some(dst),
            (Morphism { .. }, _) => None,
            (Functor { scat, .. }, 0) => Some(scat),
            (Functor { dcat, .. }, 1) => Some(dcat),
            (Functor { .. }, _) => None,
            (Equality { cat, .. }, 0) => Some(cat),
            (Equality { src, .. }, 1) => Some(src),
            (Equality { dst, .. }, 2) => Some(dst),
            (Equality { left, .. }, 3) => Some(left),
            (Equality { right, .. }, 4) => Some(right),
            (Equality { .. }, _) => None,
            (AppliedFunctObj { scat, .. }, 0) => Some(scat),
            (AppliedFunctObj { dcat, .. }, 1) => Some(dcat),
            (AppliedFunctObj { funct, .. }, 2) => Some(funct),
            (AppliedFunctObj { obj, .. }, 3) => Some(obj),
            (AppliedFunctObj { .. }, _) => None,
            (Identity { cat, .. }, 0) => Some(cat),
            (Identity { obj, .. }, 1) => Some(obj),
            (Identity { .. }, _) => None,
            (ComposeMph { cat, .. }, 0) => Some(cat),
            (ComposeMph { src, .. }, 1) => Some(src),
            (ComposeMph { mid, .. }, 2) => Some(mid),
            (ComposeMph { dst, .. }, 3) => Some(dst),
            (ComposeMph { m1, .. }, 4) => Some(m1),
            (ComposeMph { m2, .. }, 5) => Some(m2),
            (ComposeMph { .. }, _) => None,
            (AppliedFunctMph { scat, .. }, 0) => Some(scat),
            (AppliedFunctMph { dcat, .. }, 1) => Some(dcat),
            (AppliedFunctMph { funct, .. }, 2) => Some(funct),
            (AppliedFunctMph { src, .. }, 3) => Some(src),
            (AppliedFunctMph { dst, .. }, 4) => Some(dst),
            (AppliedFunctMph { mph, .. }, 5) => Some(mph),
            (AppliedFunctMph { .. }, _) => None,
            (Reflexivity { cat, .. }, 0) => Some(cat),
            (Reflexivity { src, .. }, 1) => Some(src),
            (Reflexivity { dst, .. }, 2) => Some(dst),
            (Reflexivity { mph, .. }, 3) => Some(mph),
            (Reflexivity { .. }, _) => None,
            (Concat { cat, .. }, 0) => Some(cat),
            (Concat { src, .. }, 1) => Some(src),
            (Concat { dst, .. }, 2) => Some(dst),
            (Concat { left, .. }, 3) => Some(left),
            (Concat { mid, .. }, 4) => Some(mid),
            (Concat { right, .. }, 5) => Some(right),
            (Concat { eq1, .. }, 6) => Some(eq1),
            (Concat { eq2, .. }, 7) => Some(eq2),
            (Concat { .. }, _) => None,
            (InverseEq { cat, .. }, 0) => Some(cat),
            (InverseEq { src, .. }, 1) => Some(src),
            (InverseEq { dst, .. }, 2) => Some(dst),
            (InverseEq { left, .. }, 3) => Some(left),
            (InverseEq { right, .. }, 4) => Some(right),
            (InverseEq { eq, .. }, 5) => Some(eq),
            (InverseEq { .. }, _) => None,
            (ComposeEq { cat, .. }, 0) => Some(cat),
            (ComposeEq { src, .. }, 1) => Some(src),
            (ComposeEq { mid, .. }, 2) => Some(mid),
            (ComposeEq { dst, .. }, 3) => Some(dst),
            (ComposeEq { left1, .. }, 4) => Some(left1),
            (ComposeEq { right1, .. }, 5) => Some(right1),
            (ComposeEq { eq1, .. }, 6) => Some(eq1),
            (ComposeEq { left2, .. }, 7) => Some(left2),
            (ComposeEq { right2, .. }, 8) => Some(right2),
            (ComposeEq { eq2, .. }, 9) => Some(eq2),
            (ComposeEq { .. }, _) => None,
            (Associativity { cat, .. }, 0) => Some(cat),
            (Associativity { src, .. }, 1) => Some(src),
            (Associativity { mid1, .. }, 2) => Some(mid1),
            (Associativity { mid2, .. }, 3) => Some(mid2),
            (Associativity { dst, .. }, 4) => Some(dst),
            (Associativity { m1, .. }, 5) => Some(m1),
            (Associativity { m2, .. }, 6) => Some(m2),
            (Associativity { m3, .. }, 7) => Some(m3),
            (Associativity { .. }, _) => None,
            (LeftUnitality { cat, .. }, 0) => Some(cat),
            (LeftUnitality { src, .. }, 1) => Some(src),
            (LeftUnitality { dst, .. }, 2) => Some(dst),
            (LeftUnitality { mph, .. }, 3) => Some(mph),
            (LeftUnitality { .. }, _) => None,
            (RightUnitality { cat, .. }, 0) => Some(cat),
            (RightUnitality { src, .. }, 1) => Some(src),
            (RightUnitality { dst, .. }, 2) => Some(dst),
            (RightUnitality { mph, .. }, 3) => Some(mph),
            (RightUnitality { .. }, _) => None,
            (LeftApplication { cat, .. }, 0) => Some(cat),
            (LeftApplication { src, .. }, 1) => Some(src),
            (LeftApplication { mid, .. }, 2) => Some(mid),
            (LeftApplication { dst, .. }, 3) => Some(dst),
            (LeftApplication { mph, .. }, 4) => Some(mph),
            (LeftApplication { left, .. }, 5) => Some(left),
            (LeftApplication { right, .. }, 6) => Some(right),
            (LeftApplication { eq, .. }, 7) => Some(eq),
            (LeftApplication { .. }, _) => None,
            (RightApplication { cat, .. }, 0) => Some(cat),
            (RightApplication { src, .. }, 1) => Some(src),
            (RightApplication { mid, .. }, 2) => Some(mid),
            (RightApplication { dst, .. }, 3) => Some(dst),
            (RightApplication { left, .. }, 4) => Some(left),
            (RightApplication { right, .. }, 5) => Some(right),
            (RightApplication { eq, .. }, 6) => Some(eq),
            (RightApplication { mph, .. }, 7) => Some(mph),
            (RightApplication { .. }, _) => None,
            (FunctIdentity { scat, .. }, 0) => Some(scat),
            (FunctIdentity { dcat, .. }, 1) => Some(dcat),
            (FunctIdentity { funct, .. }, 2) => Some(funct),
            (FunctIdentity { obj, .. }, 3) => Some(obj),
            (FunctIdentity { .. }, _) => None,
            (FunctComposition { scat, .. }, 0) => Some(scat),
            (FunctComposition { dcat, .. }, 1) => Some(dcat),
            (FunctComposition { funct, .. }, 2) => Some(funct),
            (FunctComposition { src, .. }, 3) => Some(src),
            (FunctComposition { mid, .. }, 4) => Some(mid),
            (FunctComposition { dst, .. }, 5) => Some(dst),
            (FunctComposition { m1, .. }, 6) => Some(m1),
            (FunctComposition { m2, .. }, 7) => Some(m2),
            (FunctComposition { .. }, _) => None,
            (AppliedFunctEq { scat, .. }, 0) => Some(scat),
            (AppliedFunctEq { dcat, .. }, 1) => Some(dcat),
            (AppliedFunctEq { funct, .. }, 2) => Some(funct),
            (AppliedFunctEq { src, .. }, 3) => Some(src),
            (AppliedFunctEq { dst, .. }, 4) => Some(dst),
            (AppliedFunctEq { left, .. }, 5) => Some(left),
            (AppliedFunctEq { right, .. }, 6) => Some(right),
            (AppliedFunctEq { eq, .. }, 7) => Some(eq),
            (AppliedFunctEq { .. }, _) => None,
        };
        self.pos += 1;
        r.copied()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        use Feature::*;
        match self.feat {
            Category => (0, Some(0)),
            Object { .. } => (1, Some(1)),
            Morphism { .. } => (3, Some(3)),
            Functor { .. } => (2, Some(2)),
            Equality { .. } => (5, Some(5)),
            AppliedFunctObj { .. } => (4, Some(4)),
            Identity { .. } => (2, Some(2)),
            ComposeMph { .. } => (6, Some(6)),
            AppliedFunctMph { .. } => (6, Some(6)),
            Reflexivity { .. } => (4, Some(4)),
            Concat { .. } => (8, Some(8)),
            InverseEq { .. } => (6, Some(6)),
            ComposeEq { .. } => (10, Some(10)),
            Associativity { .. } => (8, Some(8)),
            LeftUnitality { .. } => (4, Some(4)),
            RightUnitality { .. } => (4, Some(4)),
            LeftApplication { .. } => (8, Some(8)),
            RightApplication { .. } => (8, Some(8)),
            FunctIdentity { .. } => (4, Some(4)),
            FunctComposition { .. } => (8, Some(8)),
            AppliedFunctEq { .. } => (8, Some(8)),
        }
    }
}
impl<'a> ExactSizeIterator for FeatureIterator<'a> {}

impl Feature {
    pub fn tag(&self) -> Tag {
        use Feature::*;
        match self {
            Category => Tag::Category,
            Object { .. } => Tag::Object,
            Morphism { .. } => Tag::Morphism,
            Functor { .. } => Tag::Functor,
            Equality { .. } => Tag::Equality,
            AppliedFunctObj { .. } => Tag::AppliedFunctObj,
            Identity { .. } => Tag::Identity,
            ComposeMph { .. } => Tag::ComposeMph,
            AppliedFunctMph { .. } => Tag::AppliedFunctMph,
            Reflexivity { .. } => Tag::Reflexivity,
            Concat { .. } => Tag::Concat,
            InverseEq { .. } => Tag::InverseEq,
            ComposeEq { .. } => Tag::ComposeEq,
            Associativity { .. } => Tag::Associativity,
            LeftUnitality { .. } => Tag::LeftUnitality,
            RightUnitality { .. } => Tag::RightUnitality,
            LeftApplication { .. } => Tag::LeftApplication,
            RightApplication { .. } => Tag::RightApplication,
            FunctIdentity { .. } => Tag::FunctIdentity,
            FunctComposition { .. } => Tag::FunctComposition,
            AppliedFunctEq { .. } => Tag::AppliedFunctEq,
        }
    }

    pub fn iter<'a>(&'a self) -> FeatureIterator<'a> {
        FeatureIterator { feat: self, pos: 0 }
    }

    pub fn from_iter<I>(tag: Tag, mut iter: I) -> Option<Self>
    where
        I: Iterator<Item = u64>,
    {
        use Feature::*;
        match tag {
            Tag::Category => {
                if iter.next().is_some() {
                    None
                } else {
                    Some(Category)
                }
            }
            Tag::Object => {
                let cat = iter.exactly_one().ok()?;
                Some(Object { cat })
            }
            Tag::Morphism => {
                let (cat, src, dst) = iter.collect_tuple()?;
                Some(Morphism { cat, src, dst })
            }
            Tag::Functor => {
                let (scat, dcat) = iter.collect_tuple()?;
                Some(Functor { scat, dcat })
            }
            Tag::Equality => {
                let (cat, src, dst, left, right) = iter.collect_tuple()?;
                Some(Equality {
                    cat,
                    src,
                    dst,
                    left,
                    right,
                })
            }
            Tag::AppliedFunctObj => {
                let (scat, dcat, funct, obj) = iter.collect_tuple()?;
                Some(AppliedFunctObj {
                    scat,
                    dcat,
                    funct,
                    obj,
                })
            }
            Tag::Identity => {
                let (cat, obj) = iter.collect_tuple()?;
                Some(Identity { cat, obj })
            }
            Tag::ComposeMph => {
                let (cat, src, mid, dst, m1, m2) = iter.collect_tuple()?;
                Some(ComposeMph {
                    cat,
                    src,
                    mid,
                    dst,
                    m1,
                    m2,
                })
            }
            Tag::AppliedFunctMph => {
                let (scat, dcat, funct, src, dst, mph) = iter.collect_tuple()?;
                Some(AppliedFunctMph {
                    scat,
                    dcat,
                    funct,
                    src,
                    dst,
                    mph,
                })
            }
            Tag::Reflexivity => {
                let (cat, src, dst, mph) = iter.collect_tuple()?;
                Some(Reflexivity { cat, src, dst, mph })
            }
            Tag::Concat => {
                let (cat, src, dst, left, mid, right, eq1, eq2) = iter.collect_tuple()?;
                Some(Concat {
                    cat,
                    src,
                    dst,
                    left,
                    mid,
                    right,
                    eq1,
                    eq2,
                })
            }
            Tag::InverseEq => {
                let (cat, src, dst, left, right, eq) = iter.collect_tuple()?;
                Some(InverseEq {
                    cat,
                    src,
                    dst,
                    left,
                    right,
                    eq,
                })
            }
            Tag::ComposeEq => {
                let (cat, src, mid, dst, left1, right1, eq1, left2, right2, eq2) =
                    iter.collect_tuple()?;
                Some(ComposeEq {
                    cat,
                    src,
                    mid,
                    dst,
                    left1,
                    right1,
                    eq1,
                    left2,
                    right2,
                    eq2,
                })
            }
            Tag::Associativity => {
                let (cat, src, mid1, mid2, dst, m1, m2, m3) = iter.collect_tuple()?;
                Some(Associativity {
                    cat,
                    src,
                    mid1,
                    mid2,
                    dst,
                    m1,
                    m2,
                    m3,
                })
            }
            Tag::LeftUnitality => {
                let (cat, src, dst, mph) = iter.collect_tuple()?;
                Some(LeftUnitality { cat, src, dst, mph })
            }
            Tag::RightUnitality => {
                let (cat, src, dst, mph) = iter.collect_tuple()?;
                Some(RightUnitality { cat, src, dst, mph })
            }
            Tag::LeftApplication => {
                let (cat, src, mid, dst, mph, left, right, eq) = iter.collect_tuple()?;
                Some(LeftApplication {
                    cat,
                    src,
                    mid,
                    dst,
                    mph,
                    left,
                    right,
                    eq,
                })
            }
            Tag::RightApplication => {
                let (cat, src, mid, dst, left, right, eq, mph) = iter.collect_tuple()?;
                Some(RightApplication {
                    cat,
                    src,
                    mid,
                    dst,
                    left,
                    right,
                    eq,
                    mph,
                })
            }
            Tag::FunctIdentity => {
                let (scat, dcat, funct, obj) = iter.collect_tuple()?;
                Some(FunctIdentity {
                    scat,
                    dcat,
                    funct,
                    obj,
                })
            }
            Tag::FunctComposition => {
                let (scat, dcat, funct, src, mid, dst, m1, m2) = iter.collect_tuple()?;
                Some(FunctComposition {
                    scat,
                    dcat,
                    funct,
                    src,
                    mid,
                    dst,
                    m1,
                    m2,
                })
            }
            Tag::AppliedFunctEq => {
                let (scat, dcat, funct, src, dst, left, right, eq) = iter.collect_tuple()?;
                Some(AppliedFunctEq {
                    scat,
                    dcat,
                    funct,
                    src,
                    dst,
                    left,
                    right,
                    eq,
                })
            }
        }
    }
}
